File: src/main/java/kitchenpos/eatinorders/tobe/application/EatInOrderTableService.java
package kitchenpos.eatinorders.tobe.application;

import kitchenpos.eatinorders.tobe.domain.exception.InvalidEatInOrderTableException;
import kitchenpos.eatinorders.tobe.domain.order.EatInOrderRepository;
import kitchenpos.eatinorders.tobe.domain.orderTable.EatInOrderTable;
import kitchenpos.eatinorders.tobe.domain.orderTable.EatInOrderTableRepository;
import kitchenpos.eatinorders.tobe.domain.orderTable.OrderTableOrders;
import kitchenpos.eatinorders.tobe.domain.orderTable.vo.NumberOfGuests;
import kitchenpos.eatinorders.tobe.ui.dto.ChangeNumberOfGuestsRequest;
import kitchenpos.eatinorders.tobe.ui.dto.CreateEatInOrderTableRequest;
import kitchenpos.eatinorders.tobe.ui.dto.CreateEatInOrderTableResponse;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.NoSuchElementException;
import java.util.UUID;

@Transactional
@Service
public class EatInOrderTableService {

    private final EatInOrderRepository eatInOrderRepository;
    private final EatInOrderTableRepository eatInOrderTableRepository;
    private final OrderTableOrders orderTableOrders;

    public EatInOrderTableService(final EatInOrderRepository eatInOrderRepository,
                                  final EatInOrderTableRepository eatInOrderTableRepository,
                                  final OrderTableOrders orderTableOrders) {
        this.eatInOrderRepository = eatInOrderRepository;
        this.eatInOrderTableRepository = eatInOrderTableRepository;
        this.orderTableOrders = orderTableOrders;
    }

    public CreateEatInOrderTableResponse create(final CreateEatInOrderTableRequest request) {
        final EatInOrderTable eatInOrderTable = eatInOrderTableRepository.save(
                new EatInOrderTable(request.name()));
        return CreateEatInOrderTableResponse.from(eatInOrderTable);
    }

    public EatInOrderTable sit(final UUID orderTableId) {
        final EatInOrderTable eatInOrderTable = eatInOrderTableRepository.findById(orderTableId)
                .orElseThrow(() -> new InvalidEatInOrderTableException("주문 테이블이 존재해야 합니다."));
        eatInOrderTable.sit();
        return eatInOrderTable;
    }

    /**
     * 응용 서비스에서 유효성 검사를 도메인 서비스로 이동
     */
    public EatInOrderTable clear(final UUID orderTableId) {
        final EatInOrderTable eatInOrderTable = eatInOrderTableRepository.findById(orderTableId)
                .orElseThrow(() -> new InvalidEatInOrderTableException("주문 테이블이 존재해야 합니다."));

        eatInOrderTable.clear(orderTableOrders);
        return eatInOrderTable;
    }

    public EatInOrderTable changeNumberOfGuests(final UUID orderTableId, final ChangeNumberOfGuestsRequest request) {
        final NumberOfGuests numberOfGuests  = new NumberOfGuests(request.numberOfGuests());
        final EatInOrderTable eatInOrderTable = eatInOrderTableRepository.findById(orderTableId)
                .orElseThrow(NoSuchElementException::new);
        eatInOrderTable.changeNumberOfGuests(numberOfGuests.getNumberOfGuests());
        return eatInOrderTable;
    }
}

File: src/main/java/kitchenpos/eatinorders/tobe/application/EatInOrderTableService.kt
package kitchenpos.eatinorders.tobe.application

import kitchenpos.eatinorders.tobe.domain.*
import kitchenpos.eatinorders.tobe.dto.`in`.OrderTableCreateRequest
import kitchenpos.eatinorders.tobe.dto.out.OrderTableResponse
import kitchenpos.eatinorders.tobe.dto.out.fromOrderTable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.UUID

@Service
class EatInOrderTableService(
    private val orderTableRepository: OrderTableRepository,
    private val orderTableClearValidator: OrderTableClearValidator
) {
    @Transactional
    fun createOrderTable(orderTableCreateRequest: OrderTableCreateRequest): OrderTableResponse {
        val orderTable = OrderTable.of(OrderTableName(orderTableCreateRequest.name))

        return fromOrderTable(orderTableRepository.save(orderTable))
    }

    @Transactional
    fun sit(orderTableId: UUID): OrderTableResponse {
        val orderTable = orderTableRepository.findByIdOrNull(orderTableId)
            ?: throw NoSuchElementException("can not found order table: $orderTableId")

        orderTable.sit()

        return fromOrderTable(orderTable)
    }

    @Transactional
    fun clear(orderTableId: UUID): OrderTableResponse {
        val orderTable = orderTableRepository.findByIdOrNull(orderTableId)
            ?: throw NoSuchElementException("can not found order table: $orderTableId")

        orderTable.clear(orderTableClearValidator)

        return fromOrderTable(orderTable)
    }

    @Transactional
    fun changeNumberOfGuests(orderTableId: UUID, numberOfGuests: Int): OrderTableResponse {
        val orderTable = orderTableRepository.findByIdOrNull(orderTableId)
            ?: throw NoSuchElementException("can not found order table: $orderTableId")

        orderTable.changeNumberOfGuest(OrderTableGuestNumber(numberOfGuests))

        return fromOrderTable(orderTable)
    }
}
